dfc/tests/regression_test.go

```diff
@@ -88,7 +93,115 @@ var (
 	failLRU = ""
 )
 
-func Test_regression(t *testing.T) {
+// Intended for a deployment with multiple targets
+// 1. Unregister target T
+// 2. Create local bucket
+// 3. PUT large amount of objects into the local bucket
+// 4. GET the objects while simultaneously re-registering the target T
+func TestGetAndReRegisterInParallel(t *testing.T) {
+	var (
+		sid                string
+		num                = 20000
+		numGetsForEachFile = 5
+		numGetErrs         int
+		maxNumGetErrs      = num * numGetsForEachFile / 10000 // 0.01% of GET requests
+		filenameCh         = make(chan string, num)
+		repFilenameCh      = make(chan repFile, num)
+		errch              = make(chan error, 100)
+		sgl                *dfc.SGLIO
+		filesize           = uint64(1024)
+		seed               = int64(111)
+		wg                 sync.WaitGroup
+	)
+	// Step 1.
+	smap := getClusterMap(httpclient, t)
+	l := len(smap.Tmap)
+	if l < 2 {
+		t.Fatalf("Must have 2 or more targets in the cluster, have only %d", l)
+	}
+	for sid = range smap.Tmap {
+		break
+	}
+	unregisterTarget(sid, t)
+	n := len(getClusterMap(httpclient, t).Tmap)
+	tlogf("Unregistered target %s: the cluster now has %d targets\n", sid, n)
+
+	// Step 2.
+	bucket := TestLocalBucketName
+	err := client.CreateLocalBucket(proxyurl, bucket)
+	if err != nil {
+		t.Fatalf("client.CreateLocalBucket failed, err = %v", err)
+	}
+
+	defer func() {
+		err = client.DestroyLocalBucket(proxyurl, bucket)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+
+	if usingSG {
+		sgl = dfc.NewSGLIO(filesize)
+		defer sgl.Free()
+	}
+
+	// Step 3.
+	tlogf("Putting %d files into bucket %s...\n", num, bucket)
+	putRandomFiles(0, seed, filesize, num, bucket, t, nil, errch, filenameCh, SmokeDir, SmokeStr, "", true, sgl)
+	selectErr(errch, "put", t, false)
+	close(filenameCh)
+
+	for f := range filenameCh {
+		repFilenameCh <- repFile{repetitions: numGetsForEachFile, filename: f}
+	}
+
+	// Step 4.
+	tlogf("Getting each of the %d files %d times from bucket %s...\n", num, numGetsForEachFile, bucket)
+	wg.Add(num * numGetsForEachFile)
+	go func() {
+		// The GET is timed such that around half of all objects will be requested after target T is re-registered
+		time.Sleep(15 * time.Second)
+		for i := 0; i < num*numGetsForEachFile; i++ {
+			if (i+1)%5000 == 0 {
+				tlogf("Doing GET request %d/%d...\n", i+1, num*numGetsForEachFile)
+			}
+			go func() {
+				repFile := <-repFilenameCh
+				if repFile.repetitions > 0 {
+					repFile.repetitions -= 1
+					repFilenameCh <- repFile
+				}
+				_, _, err := client.Get(proxyurl, bucket, SmokeStr+"/"+repFile.filename, nil, nil, false, false)
+				if err != nil {
+					numGetErrs += 1
+				}
+				wg.Done()
```

Comment:
this is ok, but good practice is use defer wg.done() at the top of the function in case there are returns in the middle